EVA-1573 — Implement variant annotation pipeline in-house

tskir opened the pull request:
* Implemented the consequence mapping pipeline (see the main module [**here**](https://github.com/tskir/eva-cttv-pipeline/blob/eva-1573-functional-annotation/bin/consequence_mapping/consequence_mapping.py) with code highlighting). The pipeline includes several important improvements compared to the original Open Targets Perl pipeline (see detailed description in the documentation):
  + It uses only VCF-compatible full variant identifiers (chromosome, position, ref, alt). All other types of referring to variants, including rsIDs, have now been deprecated because they are not allele specific.
  + Upstream and downstream gene variants are now fixed in a unified way.
  + Fixed a serious bug present in the original pipeline, which led to disappearance of some variants from the output.
  + The pipeline runs in parallel, and VEP is queried with multiple variants at once, improving the efficiency.
* Updated the main code for compatibility with the new consequence mapping pipeline:
  + The most important update is that pairing between ClinVar records and functional consequence variants is now **not** attemped using rsIDs and nsvIDs, as they are not allele specific.
  + Also, now the pipeline always uses VCF-compatible positions/reference/alternate alleles.
* Documentation:
  + Wrote documentation for the new consequence mapping pipeline (see [**here**](https://github.com/tskir/eva-cttv-pipeline/blob/eva-1573-functional-annotation/docs/consequence-mapping.md) for a rendered version)
  + Updated batch submission instructions to use the new pipeline.
  + Steps 1, 2, 3 has now been merged into a single Step 1. The total number of steps is down from 7 to 5.
* Tests:
  + Updated examples for ClinVar record (it was obsolete by 5 years), consequence mappings & evidence strings
  + Updated tests for ClinVar records and consequence mappings to account for the updated data
  + Refactored tests for evidence string generation. The old tests were quite ugly, had multiple long hardcoded literals all over the place, and in some places compared the generated evidence strings with themselves (no correct “source of truth” comparison). All of this has been now rectified.
* Removed several unused modules & test files.

sundarvenkata-EBI commented on bin/consequence_mapping/consequence_mapping.py:
```diff
@@ -0,0 +1,154 @@
+
+logging.basicConfig()
+logger = logging.getLogger('consequence_mapping')
+logger.setLevel(logging.INFO)
```
Suggest moving logging initialization to __init__.py so that it can be shared across other python source files in the directory (should there be any in the future)

sundarvenkata-EBI commented on bin/consequence_mapping/run_consequence_mapping.sh:
```diff
@@ -0,0 +1,11 @@
+
+bcftools query "$1" -f '%CHROM:%POS:%REF:%ALT\n' \
+  | sort -u \
+  | parallel --pipe -j 20 -N 200 python $DIR/consequence_mapping.py \
```
Could you please consult Ensembl VEP devs to ensure that they don't have any concurrency rate limiting implemented for their web service (like we do for ours)?

sundarvenkata-EBI replied:
I am not well-versed with the failure semantics for parallel... If one of the concurrent python processes fails midway, are those outputs discarded or the pipe completely fails? the python program uses logger which prints out info and error messages... will all of those stdout/stderr outputs (presumably created in parallel) be captured coherently without race conditions? Also, I am curious if you had a chance to test the shell script by invoking it with bsub...

tskir replied:
To answer your questions:
1. I already consulted [the relevant documentation](https://github.com/Ensembl/ensembl-rest/wiki#example-clients) concerning rate limiting. They allow submitting 15 requests per second. Given that each instance of a `consequence_mapping.py` script submits only a single requests, and it takes on average 10 seconds to process, the effective request rate for 20 workers running in parallel is about 2 requests per second—a great deal less than the allowed limit.
2. Internally, `consequence_mapping.py` script catches request errors using the `@retry` decorator and attempts to retry them up to 10 times, with increasing intervals (plus some jitter). If after all attempts a request was still unsuccessful, the script will fail with non-zero return code.
3. If at least one process handled by GNU Parallel fails, the entire run will fail with non-zero return code. This is expected behaviour, because if at least chunk of data was not processed, the entire run should fail.
4. Concerning stream redirection:
  + STDOUT of `consequence_mapping.py` is redirected by `run_consequence_mapping.sh` into a specific file and is not further processed by Parallel or bsub.
  + STDERR of `consequence_mapping.py` flows through `run_consequence_mapping.sh` unchanged, is caught by Parallel (and further by bsub, if it is used).
5. I have performed end-to-end testing and checked the assertions in points 1-4 work—the Ensembl API errors are being handled and STDOUT/STDERR/script output are being correctly handled by bsub

sundarvenkata-EBI replied:
Given that you have performed end-to-end testing already, could you include running the shell script as part of the test suite too? Parallelism is notoriously hard to get right and since it looks like we have gotten it right now, having safeguards at the test suite will guard against future regressions...

tskir replied:
Yes, to have at least an integration test is a good idea. I will implement this in a separate ticket, since we will be moving this code into a separate repository and this will require a separate task anyway.

sundarvenkata-EBI commented on bin/consequence_mapping/consequence_mapping.py:
```diff
@@ -0,0 +1,154 @@
+    return [element for element, _ in itertools.groupby(sorted(lst))]
+
+
+def colon_to_vep(colon_id):
```
Feel free to expand the function name to **colon_based_id_to_vep_id**. Doesn't cost us anything :D

tskir replied:
Done, also the same for `vep_to_colon` function

sundarvenkata-EBI commented on bin/consequence_mapping/consequence_mapping.py:
```diff
@@ -0,0 +1,154 @@
+    severity_ranking = open(severity_ranking_path).read().splitlines()
+    consequence_term_severity_rank = {term: index for index, term in enumerate(severity_ranking)}
+
+    variants_to_query = [colon_to_vep(v) for v in sys.stdin.read().splitlines()]
```
Would suggest making it explicit in a comment that this program expects data from a pipe...

tskir replied:
Done

sundarvenkata-EBI commented on bin/consequence_mapping/consequence_mapping.py:
```diff
@@ -0,0 +1,154 @@
+
+def vep_to_colon(vep_id):
+    """Converts a VEP compatible identifier to colon-based."""
+    vep_id_fields = vep_id.split(' ')
```
Is there a strict specification for VEP IDs? Is the single white space requirement strictly adhered to in all the IDs?

tskir replied:
VEP support several types of variant IDs. This function converts only a single specific type of identifier (because we only query VEP with that one type of identifier). I expanded function docstring to explain that.

sundarvenkata-EBI commented on bin/consequence_mapping/consequence_mapping.py:
```diff
@@ -0,0 +1,154 @@
+    # First, we query VEP with default parameters, looking for variants affecting protein coding and miRNA transcripts
+    # up to a standard distance (5000 nucleotides either way, which is default for VEP) from the variant.
+    results_by_variant = {}
+    vep_results = query_vep(variants=variants, search_distance=5000)
```
Could we use module-level constants for numbers like this?

tskir replied:
Moved the two constants to top level, and also added description

sundarvenkata-EBI commented on bin/consequence_mapping/severity_ranking.txt:
```diff
@@ -0,0 +1,36 @@
+feature_elongation
+regulatory_region_variant
+feature_truncation
+intergenic_variant
```
Have you checked with Ensembl VEP devs if this list is something that could be retrieved with a web service (so that we don't need to have this hard-coded text file)?

tskir replied:
There is an endpoint to retrieve all consequence types: https://rest.ensembl.org/documentation/info/variation_consequence_types

Unfortunately, it returns the consequence types in arbitrary order and does not specify the severity rankings.

sundarvenkata-EBI commented on bin/consequence_mapping/consequence_mapping.py:
```diff
@@ -0,0 +1,154 @@
+    results_by_variant = {}
+    vep_results = query_vep(variants=variants, search_distance=5000)
+    extract_consequences(vep_results=vep_results, acceptable_biotypes={'protein_coding', 'miRNA'},
+                         only_closest=False, results_by_variant=results_by_variant)
```
Judgement call: Just from a programmer POV, I am a bit antsy about functions that mutate the variables that are passed in. Could we make **extract_consequences** explicitly return the modified dictionary so that we can assign the function call result to results_by_variant again? That makes it easier to figure out the places where the variable gets modified...

tskir replied:
This is a very good point! Implemented the way you suggested

sundarvenkata-EBI commented on bin/consequence_mapping/run_consequence_mapping.sh:
```diff
@@ -0,0 +1,11 @@
+# A wrapper script for running the consequence mapping pipeline concurrently and based on a VCF input.
+# See /docs/consequence-mapping.md for detailed information.
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
```
If I had a dime for every time I have written this evil-but-necessary hack to get the working directory, I would be a millionaire...

sundarvenkata-EBI commented on bin/consequence_mapping/consequence_mapping.py:
```diff
@@ -0,0 +1,154 @@
+    variants_without_consequences = get_variants_without_consequences(results_by_variant)
+    if variants_without_consequences:
+        logger.info('After distant querying, still remaining {} variant(s) without consequences: {}'.format(
+            len(variants_without_consequences), '|'.join(variants_without_consequences)
```
Suggest newline instead of pipe for readability...

tskir replied:
Yes, exactly, it's there for easier programmatic processing of the logs. I can't replace it with newline because it would break `logging` module prefixes, although I could replace it with some other symbol or combination of symbols.

sundarvenkata-EBI replied:
On second thought, pipe looks fine... you can use the cut command with the pipe delimiter in conjunction with grep to get the list of variants...

sundarvenkata-EBI commented on bin/consequence_mapping/consequence_mapping.py:
```diff
@@ -0,0 +1,154 @@
+
+        # Keep only consequences which include the most severe consequence term; sort by increasing order of distance
+        consequences = [c for c in consequences if most_severe_consequence_term in c['consequence_terms']]
+        consequences.sort(key=lambda consequence: abs(consequence.get('distance', 0)))
```
so we are making the least possible distance as the default? Shouldn't we do it the other way round?

tskir replied:
Ah, the reason for that is that if there is no `distance` attribute in VEP results, it means that it is not applicable to the variant because it resides *inside* the gene. Hence, in this case the distance is set to 0. I added a comment with this clarification to the code.

sundarvenkata-EBI commented on tests/evidence_string_generation/test_evidence_strings.py:
```diff
@@ -13,288 +12,51 @@
+        evidence_string = json.dumps(evidence_strings.CTTVGeneticsEvidenceString(*self.test_args),
+                                     sort_keys=True, indent=2)
+        expected_evidence_string = open(config.expected_genetics_evidence_string).read()
+        self.assertEqual(evidence_string, expected_evidence_string)
```
per convention, expected value comes first...

jmmut replied:
It's not a convention, it's a JUnit API design (that always seemed unintuitive to me). The [api for python's unittest](https://docs.python.org/3.5/library/unittest.html#unittest.TestCase.assertEqual) just says "`first` and `second` should be equal". I'm not sure it brings any benefit to swap all the assertions in these files, specially when the expected parameter is prefixed with "expected_".

tskir replied:
Unittest does not enforce strict requirements in its API; however, the basic example in the official documentation uses (actual, expected) for ordering: https://docs.python.org/3/library/unittest.html#basic-example

tskir commented on bin/consequence_mapping/consequence_mapping.py:
```diff
@@ -0,0 +1,154 @@
+        # The second column, set statically to 1, is not used, and is maintained for compatibility purposes
+        print('\t'.join([vep_to_colon(variant_id), '1', gene_id, gene_symbol, consequence_term, str(distance)]))
+
+    logger.info('Successfully processed {} variants'.format(len(variants_to_query)))
```
This totally makes sense! However, in light of some recent discussions with Asier, I think the best solution here would be to move this pipeline into a completely separate repository (also see below)

tskir commented on docs/consequence-mapping.md:
```diff
@@ -0,0 +1,98 @@
+* Output format for the new pipeline is a 6 column TSV, mostly the same as for the old pipeline. Changes:
+  + Column 1 (variant identifier) is using a different, VCF-compatible notation.
+  + Column 5 (functional consequence): special consequence type “nearest_gene_five_prime_end” has been dropped and replaced by conventional “upstream_gene_variant” / “downstream_gene_variant”.
+  + Column 6 (distance from variant to gene) is now always non-negative, for both upstream and downstream gene variants. It also always denotes a distance to the gene as reported by VEP, not to the nearest gene 5' end.
```
Last Thursday I showed the pipeline to Asier and he said there is a possibility that they will migrate to our implementation of the pipeline as well. In light of this, I think it makes sense to move the code of the pipeline into a separate repository, and then we could set up a release cycle, and these notes will go into the changelog of the first released version. What do you think about this?

tcezard replied:
I'm happy with this

tskir commented on eva_cttv_pipeline/evidence_string_generation/clinvar_to_evidence_strings.py:
```diff
@@ -310,25 +310,27 @@ def skip_record(clinvar_record, clinvar_record_measure, consequence_type, allele
+        return consequence_type_dict[clinvar_record_measure.clinvar_record.accession]
+
+    # Previously, the pairing was also attempted based on rsID and nsvID. This is not reliable because of lack of allele
+    # specificity, and has been removed.
```
No, because:
* nsvIDs were never actually used for pairing (I experimentally verified earlier that there are no ClinVar records which would even reach the stage of the conditional where nsvID pairing is attempted)
* As for rsIDs: each ClinVar record which can be adequately processed will necessarily have complete variant description (chromosome, allele, ref and alt). If it does not have those, then even if it has an rsID, the entire record would be useless (because VEP would not be able to process such a variant, and the record would be discarded because of a lack of functional consequence mapping).

Before integrating this into production, there will be another JIRA ticket for running additional end-to-end testing and comparison of results between old and new pipelines (in collaboration with Asier)

tskir commented on bin/consequence_mapping/consequence_mapping.py:
```diff
@@ -0,0 +1,154 @@
+
+def colon_to_vep(colon_id):
+    """Converts a colon-based identifier to VEP compatible one. Example: '15:7237571:C:T' → '15 7237571 . C T'"""
+    return '{} {} . {} {}'.format(*colon_id.split(':'))
```
The basic contigs of GRCh38 do not use colon symbols as far as I can see: https://hgdownload-test.gi.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes

Anyway, this is still a very good point; I have added an assertion _inside_ `colon_based_id_to_vep_id` to check for this

jmmut replied:
Ok, I'm not sure exactly which chromosomes contain colons, but in the hts-spec repo there were several discussions about this. If you are interested: https://github.com/samtools/hts-specs/issues/258 and interesting counts: https://github.com/samtools/hts-specs/pull/220#issuecomment-305537180. I'm ok with an assertion that tells us if we encounter this.

tskir replied:
Ah, right, the good old HLA alleles... They indeed have a notation which uses colons. Although it looks like they are present in the assembly on the contig level, not as part of any "chromosomes". Anyway, the assertion is now there so we'll be fine for now

tskir commented:
Addressed all review comments in https://github.com/EBIvariation/eva-cttv-pipeline/pull/88/commits/7def6d090e97af7a1d89477cb8a14b9af54f93ca, ready for the next round of reviews.

jmmut commented on eva_cttv_pipeline/evidence_string_generation/clinvar_to_evidence_strings.py:
```diff
@@ -310,25 +310,27 @@ def skip_record(clinvar_record, clinvar_record_measure, consequence_type, allele
+    # If no luck, try to pair by ClinVar record accession. This is not a good option because ClinVar accessions do not
+    # necessarily correspond to single variants. This pairing is only retained for compatibility purposes to handle
+    # trinucleotide repeat expansion variants records, which are added manually using RCV identifiers. This will be
+    # removed as soon as correct handling of NT repeat variants is implemented.
```
```suggestion
    # removed as soon as correct handling of TN repeat variants is implemented.
```
?

tskir replied:
NT here stands for “nucleotide”. (Yes, I too think that this abbreviation is annoying, but it is unfortunately widely used)

tskir commented on bin/consequence_mapping/consequence_mapping.py:
```diff
@@ -0,0 +1,167 @@
+    # Load severity rankings for consequence terms
+    severity_ranking_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'severity_ranking.txt')
+    severity_ranking = open(severity_ranking_path).read().splitlines()
+    consequence_term_severity_rank = {term: index for index, term in enumerate(severity_ranking)}
```
You are correct, it is a global variable and it's not very pretty. I addressed it and actually moved the code into the `main()` function: https://github.com/EBIvariation/eva-cttv-pipeline/pull/88/commits/7dea3a08db82dc06814b8f008946b076a2359f68
